### ISSUE TITLE
Stuttering audio left over after destroying StreamingGenerators

### ISSUE THREAD

**Issue by chrisnorman7** [synthizer/synthizer#46]
I know I reported this before in #36, but it seems to be back.

I've tried to produce minimal code, but it's not crashing. That makes me wonder if it's something I'm doing.

This is outside my test suite. I destroy the streaming generator - and the attached source, and there's still stuttering audio.

This seemingly only happens if I've played another sound after starting the stream, so I thought I could write that up and it would fail nicely.

The only thing that makes me think that perhaps it's not me, is that when I start streaming the same sound file again through a source as the same coordinates, the left overstutter goes away.

What information do you need from me to track this down? I can keep trying to write minimal code if that would be helpful, unless there's something else you'd suggest?

**Comment by camlorn** [comment]
Ok, so what this probably means is there's an internal buffer somewhere that needs to be zeroed that isn't getting zeroed.  Let's be annoyingly exhaustive and run through the list of questions that might possibly indicate which buffer it is.

There's been a whole series of little library updates lately.  What does `pip show synthizer` say?  Please don't upgrade before answering that on the off chance that you're a little bit out of date.

Was #36 BufferGenerator or StreamingGenerator?  I'm assuming you've seen this with multiple generator types.

When you say streaming through a source at the same coordinates, which of the following do you mean:

- Same source and same generator.
- Same source, but different generator
- Different source, but same generator
- Both different

Are you toggling panner strategies after the source has generators connected?

Do the generators in question pause correctly?

If you do something like this, does it still happen?:

```python
g1 = # a generator
g2 = # a generator
g2.gain = 0.0
src.add_generator(g1)
src.add_generator(g2)
src.remove_generator(g1)
```

Feel free to modify that to more closely match your example, but the order of the generator add/remove calls is significant-ish as is the fact that g2 is still connected (setting the gain to 0 silences).

I haven't yet tried to reproduce the StreamingGenerator case, but #36 was fixed by a library upgrade so it's probably just one of those things where there's a difference between machines un-hiding a C++ bug.  I will give thought to what might be done about becoming more robust to this sort of issue in the future.

Might be a bit slower on getting back to this.  Not sure how much Synthizer coding I'll have the bandwidth for this weekend.  We shall see.

**Comment by camlorn** [comment]
Actually, I may know what this is.  Would still be helpful for you to answer as many of the questions as you can, but I think the bug is that if you create two sources configured with HRTF panning and destroy the first one then you get audio artifacts.  I will check into that tomorrow--this may be a very quick fix.

**Comment by chrisnorman7** [comment]
OK, so, pip show synthizer shows:

```bash
Name: synthizer
Version: 0.8.5
Summary: UNKNOWN
Home-page: UNKNOWN
Author: UNKNOWN
Author-email: UNKNOWN
License: UNKNOWN
Location: c:\users\chris\src\earwax\env\lib\site-packages
Requires:
Required-by: synthizer-fx, earwax
```

I think #36 started with buffer generators, but I couldn't swear to that.
There's been lots of Earwax changes, and at some point I switched over
ambiances from buffers to streams, but I don't remember when that was in
relation to the actual bug.

I'm talking about both source and generator being different.

Actually, I'm not changing the panner strategy at all. Glad you reminded me
of that - I need to set to HRTF.

They pause and play fine, yes. Interestingly, if I pause them before
stopping them, nothing, otherwise I get the buzzing. Also didn't realise
there was now a play and a pause method. I'm going to go over the manual
with a fine toothed comb when 1.0 drops.

So in attempting to put your example into my code, I found a bug of my own,
which inadvertently fixed the issue at my end: I wasn't destroying the
source; I forgot to put in the brackets.

So in short, I guess it works, as long as I remember to destroy all sources.

No worries about not coding much. I hope you have a good
weekend, regardless of what you get up to.

Take care,

Chris Norman

**Comment by camlorn** [comment]
I'm 99% sure it doesn't work, but that the current adopters are doing stuff in such a way as to hide the bug. More news when there's more news.

**Comment by chrisnorman7** [comment]
No worries mate.

Thanks for all your hard work.

Take care,

Chris Norman

**Comment by camlorn** [comment]
I'll have this in the next release, which will probably be sometime this weekend.

**Comment by chrisnorman7** [comment]
Great, thanks.

Take care,

Chris Norman
